File: hard/1373-max-sum-BST-in-BT/sol.cpp
using namespace std;

class Solution {
private:
    int* postorder(TreeNode* ptr, int &maxSum) {
        // return type: [ min, max, curr_sum ]
        if(!ptr) return new int[3]{ INT_MAX, INT_MIN, 0 };

        int currSum = INT_MIN;
        int *l  = postorder(ptr->left, maxSum);
        int *r = postorder(ptr->right, maxSum);
        int *res = new int[3]{ INT_MIN, INT_MAX, 0 };

        // maxLeft < node < minRight
        if(ptr->val < *r && ptr->val > *(l+1)) {
            currSum = *(l+2)+*(r+2)+ptr->val;
            res[0] = min(ptr->val,*l);
            res[1] = max(ptr->val,*(r+1));
        }

        if(currSum > INT_MIN) {
            res[2] = currSum;
            maxSum = max(maxSum, currSum);
        }

        delete[] l,r;
        return res; 
    }

public:
    int maxSumBST(TreeNode* root) {
        int maxSum = INT_MIN;
        postorder(root,maxSum);
        return max(0,maxSum);
    }
};

File: easy/0653-2-sum-IV-input-is-BST/sol.py
class Solution:
    def findTarget(self, root: Optional[TreeNode], k: int) -> bool:
        
        def inorder(ptr):
            nonlocal isFound, wanted
            if not ptr or isFound: return
            if ptr.val in wanted:
                isFound = True
                return
            
            wanted.add(k-ptr.val)

            if not isFound: inorder(ptr.left)
            if not isFound: inorder(ptr.right)

        
        wanted = set()
        isFound = False
        inorder(root)

        return isFound